### ISSUE TITLE
Infinite loop

### ISSUE THREAD

**Issue by bryanforbes** [domenic/promises-unwrapping#72]
While writing a TypeScript implementation of this spec using both the spec and the testable implementation for guidance, I found an infinite loop when calling a promise's resolver with itself. The testable implementation has the following code in the reaction function returned from `make_PromiseReactionFunction`:

``` js
if (SameValue(handlerResult, deferred["[[Promise]]"]) === true) {
    let selfResolutionError = new TypeError("Tried to resolve a promise with itself!");
    Call(deferred["[[Reject]]"], selfResolutionError);
}
```

However, by the time these lines are encountered, the following has already been run in the resolution handler from `make_PromiseResolutionHandlerFunction`:

``` js
if (IsPromise(coerced)) {
    return coerced.then(fulfillmentHandler, rejectionHandler);
}
```

I thought it might just be an issue with my implementation, but I just confirmed the testable implementation has the same problem by saving the following code to `test.js` in the root of the project and running it with `node --harmony test.js`:

``` js
var impl = require('./testable-implementation');

var def = impl.deferred();

def.promise.then(
    function (value) {
        console.log('Value: ', value);
    },
    function (error) {
        console.log('Error: ', error);
    }
);

def.resolve(def.promise);
```

**Comment by kriskowal** [comment]
Thanks @bryanforbes. This issue duplicates, #223, which is closed. If you search for `ViciousCycle` in `q.js` on `master`, you’ll find how we deal with this problem now, based on @erights solution in `makeQ.js` for Caja/SES.

If you have time to do us a favor, please do verify that it works. I am going to leave this open to track “Add test case for vicious cycle detection”. It seems I neglected it.

_Edit: mistook this as an issue filed against Q. ignore…ignore_

**Comment by domenic** [comment]
Haha @kriskowal wrong repo I think :).

@bryanforbes excellent find, thank you. Perhaps related to some recent comments by @rossberg-chromium. This also means the Promises/A+ tests are not doing their job; I will file a bug over there.

**Comment by kriskowal** [comment]
@domenic Yeah, already caught and left an edit note. Sorry for the noise.

**Comment by domenic** [comment]
This would be by far easiest to fix if I was allowed to fail at resolution time. However, I am not sure this is OK with the monad contingent, as monads wrapping themselves seems valid. @rossberg-chromium, @tabatkins, @samth can you confirm one way or the other?

The behavior I am proposing is:

``` js
let resolve;
const p = new Promise(r => resolve = r);

resolve(p); // instead of setting p's [[Resolution]] to p, sets p's [[Reason]] to a TypeError.

// Thus:

p.flatMap(
  v => console.log("never called"),
  r => console.log("this is a TypeError:" + r)
);
```

Assuming the above is not OK, I will need to track more internal state and jump through a few more hoops in `then`, which is fine. But I wanted to confirm first.

**Comment by samth** [comment]
@domenic I'm confused about your code example in the last comment (why would that do anything other than try to call `undefined` as a function?) but there's no reason why these kinds of programs make any sense, and there shouldn't be any problem in disallowing them.

**Comment by domenic** [comment]
@samth `resolve` gets set to `p`'s resolve function one the second line.

**Comment by erights** [comment]
Failing at resolution time doesn't generalize to the distributed case. In the distributed case, all the local resolutions needed to form a distributed loop can happen before any of the participating machines can know they are in a loop. The current makeQ behavior is consistent with CapTP, and with what's possible in a distributed system: All the promises which come to be promises for themselves eventually become rejected as in a vicious cycle, and all messages sent into the loop are eventually dropped with the promises for their results eventually being rejected.

By contrast, there's no place to report the _eventual_ failure of a resolve.

**Comment by erights** [comment]
Ah, but I see your point regarding the monadic layering. I do not know what to do about that.

**Comment by samth** [comment]
@erights There's no problem with "monadic" layering here -- if you think about Promise<C> as a way to eventually get a C (which is the heart of what you refer to as the "monadic" position), then this issue is about promises in which there is no underlying value there, and thus doesn't even make sense.

**Comment by domenic** [comment]
I was under the impression the monad people considered array to be a monad, and `var a = []; a[0] = a` seems fine.

**Comment by samth** [comment]
Array is a monad only if you're thinking about a certain subset of its operations, which wouldn't let you construct that.

**Comment by erights** [comment]
Hi Sam, I don't understand your answer but it probably doesn't matter. As long as we're all happy with the "cyclic promise eventually becomes rejected" position, great.

**Comment by tabatkins** [comment]
> I was under the impression the monad people considered array to be a monad

We don't "consider it to be a monad", it _is_ a monad by the definition of monads (as soon as you define a flatMap() function that acts correctly), in the same way that it's an iterable by the definition of iterables - "monad" is just a typeclass, or an interface, or whatever your personal lexicon calls that kind of thing.  ("Monad" is just a more advanced form of "Mappable", the interface for anything with a reasonable .map(fn) operation, which Haskell people call Functor because they hate being understood.)

> and `var a = []; a[0] = a` seems fine.

Yes, that's fine.  The monad-ness of an array doesn't care about what's inside of it, any more than the iterable-ness of an array does.

> As long as we're all happy with the "cyclic promise eventually becomes rejected" position, great.

That's completely acceptable.

**Comment by rossberg** [comment]
[Sorry for the delay, I was on vacation.]

If we want to detect cycles then this should not happen on the resolution side, but on the chaining side, since it depends on what chaining semantics you pick (flatMap vs then). Also, it is fine for a promise to be resolved with itself, but not with its own _content_ -- an important distinction!

More specifically, I expect the following behaviour, dictated by parametricity for resolve/flatMap:

```
let deferred = Promise.deferred()
let p = deferred.promise
deferred.resolve(p)  // Not an error, p.[[value]] = p, a promise can contain itself just fine
p.flatMap(function(x) { assert(x === p) })  // Not an error
p.then(unreachable, function(r) { assert(r instanceof TypeError) })  // Error, recursive unwrapping uncovers cycle

let p1 = Promise.resolved(0)
let p2 = p1.flatMap(function(x) { return p2 })  // p2 rejected, because wants to resolve p2 with its own content (remember, flatMap does 1-level unwrapping)
let p3 = p1.then(function(x) { return p3 })  // likewise
```

None of this requires any extra state. Domenic, please see the latest check-in to my prototype, which implements this semantics with a fairly trivial change.

I'm a bit confused about some of the monad argument being made above. Just to make sure we're all on the same page, `a = [a]` is perfectly fine for a monad, provided it has a suitably recursive type. The case that corresponds to the erroneous flatMap above would be `a = [x | x <- a]`, and that would also just diverge e.g. in Haskell (as it is just a contrived way to define `a = a`).
